### ISSUE TITLE
Move mod/common.php to src/

### ISSUE THREAD

**Issue by MrPetovan** [friendica/friendica#8918]
Part of #4090

I'm having trouble navigating the common friend feature. The `/common` module has weird parameters, punitive restrictions but above all, the `Module\GContact` class it relies on has two different SQL queries to count common friends and retrieve them.

Common friends count query:
```sql
SELECT COUNT(*) as `total`
FROM `glink`
INNER JOIN `gcontact` ON `glink`.`gcid` = `gcontact`.`id`
WHERE `glink`.`cid` = %d
  AND `glink`.`uid` = %d
  AND NOT `gcontact`.`failed`
  AND `gcontact`.`nurl` IN (
    SELECT nurl
    FROM contact
    WHERE uid = ?
      AND self = 0
      AND blocked = 0
      AND hidden = 0
      AND id != ?
)
```
Common friends retrieval query:
```sql
SELECT `contact`.*
FROM `glink`
INNER JOIN `gcontact` ON `glink`.`gcid` = `gcontact`.`id`
INNER JOIN `contact` ON `gcontact`.`nurl` = `contact`.`nurl`
WHERE `glink`.`cid` = ?
  AND `glink`.`uid` = ?
  AND `contact`.`uid` = ?
  AND `contact`.`self` = 0
  AND `contact`.`blocked` = 0
  AND `contact`.`hidden` = 0
  AND `contact`.`id` != ?
  AND NOT `gcontact`.`failed`
LIMIT ?, ?
```
For some reason the queries to count common friends based on a `zcid` are identical and look like the former query.

But which one is the correct one from the two above? If they are both correct, which one is supposed to be the fastest?

@annando, do you have an idea?

**Comment by MrPetovan** [comment]
Also, what is a `zcid`? This isn't explicited in the `glink` table structure, and the only related information I could find in the comments mention a "global" contact, which I thought was the `gcid`?

**Comment by annando** [comment]
This `glink` table has some flaws in its structure. There is the table `gfollower` as a replacement. But there is still some work to do, to make it a full replacement. I can work on that after my current PRs are done.

**Comment by MrPetovan** [comment]
I’ve started separating the two module modes. `loc` (local user comparing his contacts with a contact’s relationships) and `rem` (remote visitor comparing his relationships with a local user’s contacts). Either needs only one parameter (instead of two at the moment), the contact id for `loc` which becomes `/contact/<id>/common`, the user nickname for `rem` that becomes `/profile/<nick>/common`.

I just need the common friends queries to be sorted out and I can complete this task.

**Comment by annando** [comment]
We can now distinguish between common friends ("gfollower" table) and common contacts ("contact-relation" and "gcontact-relation"). I would suggest displaying these two groups in different lists.

**Comment by MrPetovan** [comment]
Blocked by #8920

**Comment by annando** [comment]
This is now blocked by https://github.com/friendica/friendica/pull/8927

**Comment by MrPetovan** [comment]
Now that #8927 is merged what are the correct queries to fetch the common relations between a local user and an arbitrary account (contact) and between a remote visitor and a local user?

**Comment by annando** [comment]
To see the common contacts (either follower or interacting) between contact 1 and 2 you execute this query:

```
SELECT * FROM `contact` WHERE `id` IN (SELECT `cid` FROM `contact-relation` WHERE `relation-cid` = 1) AND `id` IN (SELECT `cid` FROM `contact-relation` WHERE `relation-cid` = 2);
```

To see common contacts that both contacts are following, execute this query:

```
 SELECT * FROM `contact` WHERE `id` IN (SELECT `cid` FROM `contact-relation` WHERE `relation-cid` = 1 AND `follows`) AND `id` IN (SELECT `cid` FROM `contact-relation` WHERE `relation-cid` = 2 AND `follows`);
```

To see common contacts that are following both contacts, this query should work:

```
SELECT * FROM `contact` WHERE `id` IN (SELECT `relation-cid` FROM `contact-relation` WHERE `cid` = 1 AND `follows`) AND `id` IN (SELECT `relation-cid` FROM `contact-relation` WHERE `cid` = 2 AND `follows`);
```

Common contacts that are interacting on both contacts you see with this query;

```
SELECT * FROM `contact` WHERE `id` IN (SELECT `cid` FROM `contact-relation` WHERE `relation-cid` = 1 AND `last-interaction` > '0001-01-01 00:00') AND `id` IN (SELECT `cid` FROM `contact-relation` WHERE `relation-cid` = 2 AND `last-interaction` > '0001-01-01 00:00');
```

**Comment by MrPetovan** [comment]
So I should always get a contact id from the local user (either search way) and match on contact ids, got it.

**Comment by annando** [comment]
Youu need to have the public contact id of both contacts. The nice thing is that you can even separate between the direction of the interaction on the common contacts.

**Comment by annando** [comment]
I suggest two endpoints instead of the single "common". I think we should have one for displaying who interacted with/is following the given account - and one that displays where this account interacted with/is following.

Question is if we want to separate between interaction and following. Then we would had 4 endpoints.

**Comment by MrPetovan** [comment]
I'm actually separating the view where you match against a user on a remote node when visiting or against a contact on your local node. Then it's just a matter of having an extra parameter to choose the scope of the common search.

So far I've limited it to common follows because it's the closest from the previous feature and the most intuitive. I could expand it to interactions but don't intend on doing the opposite way, random people who both interacted with both contacts. It seems like a potential for disaster since the current common queries don't take `user-contact` choices like ignore and blocks, only at the node level.

**Comment by annando** [comment]
How far are you with this? If you hadn't made any significant changes by now, I would like to replace the last functions from `GContact`. And I just realized how overcomplicated the functions are. The 4 functions can easily replace with 2 functions.

**Comment by MrPetovan** [comment]
I have the profile endpoint and the ContactRelation methods locked in. I can submit them right now and keep trudging on the rest.

**Comment by annando** [comment]
Concerning the two ways of invoking the common contacts (local and remote) and having a `$cid` and not having it:

In fact this can all be reduced to a single function call. During my experiments this weekend I discovered the function `Contact::getPublicIdByUserId` and the missing `$cid` is of course no problem anymore.

The weekend (while watching the SpaceX stream) I spent the waiting time with playing around with the "common.php", this is my result: https://github.com/annando/friendica/blob/common-contacts/mod/common.php

I guess that most of this branch is now in the other PR and since you are working on the `common.php` I wouldn't want to interfere, so I will not make a PR for this branch.
